### PR TITLE
Clean up binary build cmake script

### PR DESCRIPTION
**Pull request by Yangqing**
After the change we will be able to simply define targets and find dependencies.

Test: built locally.

### CODE REVIEW

File: caffe2/CMakeLists.txt [hunk 1]
```diff
@@ -223,20 +222,41 @@ if (BUILD_PYTHON)
...
+  # Caffe2 proto files
+  install(DIRECTORY ${CMAKE_BINARY_DIR}/caffe2 DESTINATION ${CMAKE_INSTALL_PREFIX}
+          FILES_MATCHING PATTERN "*.py")
+endif()
+
+
+# ---[ Any builds that should happen after the main targets should be added here.
+# Binaries
+if (BUILD_BINARY)
+  add_subdirectory(binaries)
```
Comment (lukeyeager) [inline caffe2/CMakeLists.txt:260]: While you're at it, is it an option to install things to `./bin/` like most other projects instead of to `./binaries/`?

Comment (Yangqing) [reply caffe2/CMakeLists.txt:260]: Yup, sounds good, will do.

Comment (QinZiwen) [reply caffe2/CMakeLists.txt:260]: so cool

Comment (aaronmarkham) [reply caffe2/CMakeLists.txt:260]: Did this change maybe result in discrepancies in where we call certain binary files from tutorials? See this comment:
https://github.com/caffe2/caffe2/issues/349#issuecomment-309502165

**Pull request closed without merging**